docs/architecture/adr-071-state-rent-unclaimed-accounts.md

```diff
@@ -0,0 +1,156 @@
+of the chain (e.g. min fees).
+
+With the advent of [#19188](https://github.com/cosmos/cosmos-sdk/pull/19188), the
+execution flow is changed such that the account is not created automatically. Instead,
+the balance is set on the recipient's address. Then as soon as the recipient sends
+their first transaction, the account is created.
+
+However, even with this improvement in execution flow, there are still DoS and
+state bloat implications. E.g. iterating over a large balance array can drastically
+degrade the performance of a chain in `BeginBlock` and `EndBlock`.
+
+Thus, we propose a mechanism for charging state rent on yet-to-be-claimed accounts/balances.
+This mechanism can further be extended and adapted to charge state rent for existing
+objects on-chain.
+
+## Decision
+
+We propose to adopt a similar mechanism to Solana's [state rent algorithm](https://docs.solanalabs.com/implemented-proposals/rent). Specifically, we propose that unclaimed account balance's be charged some
+dynamic rent, set by governance and controlled by on-chain parameters every epoch.
+
+When a user sends tokens to a non-existent account, only that account's balance
+will exist until the recipient sends their first transaction. All non-existent
+account balance's will be indexed using a collections map, indexed by the amount
+and address.
+
+We define some threshold `MinBalance`, such that all balances for non-existent
+accounts that fall below this threshold, will be charged a dynamic rent fee every
+configurable epoch, `E`. If the non-existent account's balance is no longer
+sufficient to cover the rent or the balance reaches zero, the balance will be
+permanently deleted from state. In other words, the recipient of the balance must
+send their first transaction prior to being delinquent, otherwise the recipient
+looses their funds.
+
+The collected rent can then be sent to the community pool, burned, or used in some
+combination of the two. Note, rent will be charged per unique denomination in the
+unclaimed account's balance.
+
+We define the collection index as follows, such that we only iterate over balances
+that are below the threshold `MinBalance`:
+
+```go
+var (
+  // ...
+
+  NewAccountByAmountPrefix = collections.NewPrefix(0)
+)
+
+type Keeper struct {
+  // ...
+
+  NewAccKeySet collections.KeySet[collections.Triple[sdkmath.Int, string, []byte]] // <balance, denom, address>
+}
+
+func NewKeeper(...) {
+  k := Keeper{
+    // ...
+
+    NewAccKeySet: collections.NewKeySet(
+      schemaBuilder,
+      NewAccountByAmountPrefix,
+      "new_account_by_amount",
+      collections.TripleKeyCodec(collections.IntKey, collections.StringKey, collections.BytesKey),
+    ),
+  }
+
+  // ...
+}
+```
+
+We can then define iteration as follows:
+
+```go
+k.NewAccKeySet.Walk(ctx, nil, func(key collections.Triple[sdkmath.Int, string, []byte]) (bool, error) {
+  amount := key.K1()
+  if amount.LT(MinBalance) {
+    // Charge rent...
+  }
+})
```

Comment:
minor challenge, can we do this without iteration. I think its fairly hard but seems like others do it

Reply:
Mentally, I have difficulty grasping how you "get all accounts/balances that meet condition X" w/o iterating. You'd need some loop either way.